src/sharding/Shard.js

```diff
@@ -270,33 +270,35 @@ class Shard extends EventEmitter {
    * @returns {Promise<*>} Result of the script execution
    */
   eval(script) {
+    // Stringify the script if it's a Function
+    const _eval = typeof script === 'function' ? `(${script})(this)` : script;
+
     // Shard is dead (maybe respawning), don't cache anything and error immediately
     if (!this.process && !this.worker) return Promise.reject(new Error('SHARDING_NO_CHILD_EXISTS', this.id));
 
     // Cached promise from previous call
-    if (this._evals.has(script)) return this._evals.get(script);
+    if (this._evals.has(_eval)) return this._evals.get(_eval);
```

Comment:
completely unrelated but this return should be wrapped in `Promise.resolve(...)` !

Reply:
Yeah, `Shard#_evals` is a Map of promises so `this._evals.get(_eval)` will return a Promise (maybe already resolved) of the last execution with the same code.